ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,292 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Scheduling subsystem (in-OSTD part).
+//!
+//! This module defines what OSTD expects from a scheduling implementation
+//! and provides useful functions for controlling the execution flow.
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor, task::Task, TaskStatus};
+use crate::{
+    arch::{self, timer},
+    cpu::this_cpu,
+    prelude::*,
+};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+static SHOULD_PREEMPT: ShouldPreemptFlag = ShouldPreemptFlag::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+    SCHEDULER.call_once(|| scheduler);
+
+    timer::register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                SHOULD_PREEMPT.set();
+            }
+        })
+    });
+}
+
+/// Enqueues a newly built task.
+///
+/// Note that the new task is not guranteed to run at once.
+pub(super) fn run_new_task(runnable: Arc<Task>) {
+    // FIXME: remove this check for `SCHEDULER`.
+    // Currently OSTD cannot know whether its user has injected a scheduler.
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::default());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+
+    let cpu_id_option = SCHEDULER
+        .get()
+        .unwrap()
+        .enqueue(runnable, EnqueueFlags::Spawn);
+    if cpu_id_option.is_some() {
+        let cpu_id = cpu_id_option.unwrap();
+        // FIXME: send ipi to set remote CPU's `SHOULD_PREEMPT` if needed.
```

Comment:
As I said, `SHOULD_PREEMPT` should be a temporary expedient. It would never exist if we can call `schedule` in the interrupt context. So we would never set the `SHOULD_PREEMPT` flag on remote CPUs. You can remove the `FIXME` here.

Reply:
If we can reach a agreement that kernel preemption will not be introduced to Asterinas in a near future (at least no earlier than SMP full support), then i guess we'd better keep it.

Furthermore, I suspect we'd still need this even if we finally enable kernel preemption. The basic idea in the scheduling model of this PR is that the update of scheduling statistics and the actual preemption are two separated operations and not bound to each other. We update often but only preempt on carefully chosen time points.  There are equivalents in Linux's latest release (a `TIF_NEED_RESCHED` flag in thread info and a `need_resched` flag in preempt count).